advanced/button-interrupt/examples/solution_led.rs

```diff
@@ -1,82 +1,45 @@
-
 use anyhow::Result;
 use esp_idf_svc::{
-    hal::prelude::Peripherals,
-    sys::{
-        esp, esp_random, gpio_config, gpio_config_t, gpio_install_isr_service,
-        gpio_int_type_t_GPIO_INTR_POSEDGE, gpio_isr_handler_add, gpio_mode_t_GPIO_MODE_INPUT,
-        xQueueGenericCreate, xQueueGiveFromISR, xQueueReceive, QueueHandle_t, ESP_INTR_FLAG_IRAM,
+    hal::{
+        gpio::{InterruptType, PinDriver, Pull},
+        peripherals::Peripherals,
+        task::notification::Notification,
     },
+    sys::esp_random,
 };
 use rgb_led::{RGB8, WS2812RMT};
-use std::ptr;
-
-// This `static mut` holds the queue handle we are going to get from `xQueueGenericCreate`.
-// This is unsafe, but we are careful not to enable our GPIO interrupt handler until after this value has been initialised, and then never modify it again
-static mut EVENT_QUEUE: Option<QueueHandle_t> = None;
-
-#[link_section = ".iram0.text"]
-unsafe extern "C" fn button_interrupt(_: *mut core::ffi::c_void) {
-    xQueueGiveFromISR(EVENT_QUEUE.unwrap(), std::ptr::null_mut());
-}
+use std::num::NonZeroU32;
 
 fn main() -> Result<()> {
-    let peripherals = Peripherals::take().unwrap();
+    esp_idf_svc::sys::link_patches();
 
+    let peripherals = Peripherals::take()?;
     let mut led = WS2812RMT::new(peripherals.pins.gpio2, peripherals.rmt.channel0)?;
-    const GPIO_NUM: i32 = 9;
 
     // Configures the button
-    let io_conf = gpio_config_t {
-        pin_bit_mask: 1 << GPIO_NUM,
-        mode: gpio_mode_t_GPIO_MODE_INPUT,
-        pull_up_en: true.into(),
-        pull_down_en: false.into(),
-        intr_type: gpio_int_type_t_GPIO_INTR_POSEDGE, // Positive edge trigger = button down
-    };
+    let mut button = PinDriver::input(peripherals.pins.gpio9)?;
+    button.set_pull(Pull::Up)?;
+    button.set_interrupt_type(InterruptType::PosEdge)?;
 
-    // Queue configurations
-    const QUEUE_TYPE_BASE: u8 = 0;
-    const ITEM_SIZE: u32 = 0; // We're not posting any actual data, just notifying
-    const QUEUE_SIZE: u32 = 1;
+    // Configures the notification
+    let notification = Notification::new();
+    let notifier = notification.notifier();
 
+    // Subscribe and create the callback
+    // Safety: make sure the `Notification` object is not dropped while the subscription is active
     unsafe {
-        // Writes the button configuration to the registers
-        esp!(gpio_config(&io_conf))?;
-
-        // Installs the generic GPIO interrupt handler
-        esp!(gpio_install_isr_service(ESP_INTR_FLAG_IRAM as i32))?;
-
-        // Instantiates the event queue
-        EVENT_QUEUE = Some(xQueueGenericCreate(QUEUE_SIZE, ITEM_SIZE, QUEUE_TYPE_BASE));
-
-        // Registers our function with the generic GPIO interrupt handler we installed earlier.
-        esp!(gpio_isr_handler_add(
-            GPIO_NUM,
-            Some(button_interrupt),
-            std::ptr::null_mut()
-        ))?;
+        button.subscribe(move || {
```

Comment:
NIT: to make it more clear that you can encode information inside the Notification we use `let bitset = 0b10001010101;` and send it into the NonZeroU32. You can also read out the value from `notification.wait()`

Reply:
I thought about this, but ended up removing it as it may also cause confusion since the information sent is random and completely irrelevant. But I'm happy to add it back if you think it has more value